### PR TITLE
Improved audit logging

### PR DESCRIPTION
**Pull request by danielhlarkin**

### CODE REVIEW

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/4042/

File: Documentation/Books/Manual/Security/Auditing/AuditConfiguration.md [hunk 1]
```diff
@@ -27,4 +27,16 @@ Hostname
...
+---------
+
+`--log.level topic=level`
+
+By default, the server will log all audit events. Some low-priority events, such
+as statistics operations, are logged with the `debug` log level. To keep such
+events from cluttering the log, set the appropriate topic to `info`. All other
+messages will be logged at the `info` level. Audit topics include
+`audit-authentication`, `audit-collection`, `audit-database`, `audit-documents`,
+`audit-service`, and `audit-views`.
```
Comment (Simran-B) [inline Documentation/Books/Manual/Security/Auditing/AuditConfiguration.md]: These topics don't seem to match what is written here: https://docs.arangodb.com/devel/HTTP/AdministrationAndMonitoring/index.html#modify-and-return-the-current-server-log-level

- `audit-documents` vs. `audit-documentation`
- `audit-views` vs. `audit-view`

I don't know what is correct, can you please adjust it in this file or the [other one](https://github.com/arangodb/arangodb/edit/feature/improved-audit-logging/Documentation/DocuBlocks/Rest/Administration/get_admin_modules_flush.md)?

File: arangod/RestHandler/RestAuthHandler.cpp [hunk 2]
```diff
@@ -111,3 +112,15 @@ RestStatus RestAuthHandler::badRequest() {
                 "invalid JSON");
   return RestStatus::DONE;
 }
+
+void RestAuthHandler::shutdownExecute(bool isFinalized) noexcept {
+  TRI_DEFER(RestVocbaseBaseHandler::shutdownExecute(isFinalized));
```
Comment (jsteemann) [inline arangod/RestHandler/RestAuthHandler.cpp]: I think the TRI_DEFER is unneeded here, as the events logging is encapsulated in a try catch.
So shutdownExecute could just be called after the try..catch.

File: arangod/RestHandler/RestDocumentHandler.cpp [hunk 3]
```diff
@@ -74,6 +75,30 @@ RestStatus RestDocumentHandler::execute() {
   return RestStatus::DONE;
 }
 
+void RestDocumentHandler::shutdownExecute(bool isFinalized) noexcept {
+  TRI_DEFER(RestVocbaseBaseHandler::shutdownExecute(isFinalized));
```
Comment (jsteemann) [inline arangod/RestHandler/RestDocumentHandler.cpp]: can be changed from a TRI_DEFER to a simple call of `shutdownExecute`, and can be moved behind the try..catch block.

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/4052/

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/4062/

File: Documentation/Books/Manual/Security/Auditing/AuditEvents.md [hunk 4]
```diff
@@ -117,36 +120,36 @@ Documents
 ### Reading a single document
```
Comment (Simran-B) [inline Documentation/Books/Manual/Security/Auditing/AuditEvents.md]: the headline says **Reading** but the logged message is `create document ...`?

File: Documentation/Books/Manual/Security/Auditing/AuditEvents.md [hunk 5]
```diff
@@ -117,36 +120,36 @@ Documents
...
-2016-10-04 12:28:26 | server1 | user1 | database1 | 127.0.0.1:53699 | http basic | delete document failed | /_api/document/collection1/21456?ignoreRevs=false
+2016-10-04 12:28:26 | server1 | audit-documents | user1 | database1 | 127.0.0.1:53699 | http basic | delete document 'collection/21456' | failed | /_api/document/collection1/21456?ignoreRevs=false
 ```
 
 Queries
 -------
 
 ```
-2016-10-06 12:12:10 | server1 | user1 | database1 | 127.0.0.1:54232 | http basic | query document | ok | for i in collection1 return i | /_api/cursor
+2016-10-06 12:12:10 | server1 | audit-documents | user1 | database1 | 127.0.0.1:54232 | http basic | query document | ok | for i in collection1 return i | /_api/cursor
```
Comment (Simran-B) [inline Documentation/Books/Manual/Security/Auditing/AuditEvents.md]: are the log topics in singular form (e.g. `audit-document`) but the 3rd column in the log entries in plural form (here: `audit-documents`)?

File: Documentation/Books/Manual/Security/Auditing/AuditEvents.md (hunk 4 above)
Comment (danielhlarkin) [reply Documentation/Books/Manual/Security/Auditing/AuditEvents.md]: Ah, good catch! Should be fixed now.

File: Documentation/Books/Manual/Security/Auditing/AuditEvents.md (hunk 5 above)
Comment (danielhlarkin) [reply Documentation/Books/Manual/Security/Auditing/AuditEvents.md]: Good catch.

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/4068/

**Pull request merged**